src/core/Akka.Streams/Implementation/ChannelSources.cs

```diff
@@ -0,0 +1,155 @@
+// //     Copyright (C) 2009-2022 Lightbend Inc. <http://www.lightbend.com>
+// //     Copyright (C) 2013-2022 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// // </copyright>
+// //-----------------------------------------------------------------------
+
+using System;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Akka.Streams.Stage;
+
+namespace Akka.Streams.Implementation
+{
+    sealed class ChannelSourceLogic<T> : OutGraphStageLogic
+    {
+        private readonly Outlet<T> _outlet;
+        private readonly ChannelReader<T> _reader;
+        private readonly Action<bool> _onValueRead;
+        private readonly Action<Exception> _onValueReadFailure;
+        private readonly Action<Exception> _onReaderComplete;
+        private readonly Action<Task<bool>> _onReadReady;
+
+        public ChannelSourceLogic(SourceShape<T> source, Outlet<T> outlet,
+            ChannelReader<T> reader) : base(source)
+        {
+            _outlet = outlet;
+            _reader = reader;
+            _onValueRead = GetAsyncCallback<bool>(OnValueRead);
+            _onValueReadFailure =
+                GetAsyncCallback<Exception>(OnValueReadFailure);
+            _onReaderComplete = GetAsyncCallback<Exception>(OnReaderComplete);
+            _onReadReady = ContinueAsyncRead;
+            _reader.Completion.ContinueWith(t =>
+            {
+                if (t.IsFaulted) _onReaderComplete(t.Exception);
+                else if (t.IsCanceled)
+                    _onReaderComplete(new TaskCanceledException(t));
+                else _onReaderComplete(null);
+            });
+
+            SetHandler(_outlet, this);
+        }
+
+        private void OnReaderComplete(Exception reason)
+        {
+            if (reason is null)
+                CompleteStage();
+            else
+                FailStage(reason);
+        }
+
+        private void OnValueReadFailure(Exception reason) => FailStage(reason);
+
+        private void OnValueRead(bool dataAvailable)
+        {
+            if (dataAvailable && _reader.TryRead(out var element))
+                Push(_outlet, element);
+            else
+                CompleteStage();
+        }
+
+        public override void OnPull()
+        {
+            if (_reader.TryRead(out var element))
+            {
+                Push(_outlet, element);
+            }
+            else
+            {
+                var continuation = _reader.WaitToReadAsync();
+                if (continuation.IsCompletedSuccessfully)
+                {
+                    var dataAvailable = continuation.GetAwaiter().GetResult();
+                    if (dataAvailable && _reader.TryRead(out element))
+                        Push(_outlet, element);
+                    else
+                        CompleteStage();
+                }
+                else
+                    continuation.AsTask().ContinueWith(_onReadReady);
```

Comment:
This is a slight change from the Alpakka code:

 - The Alpakka code [looked a little racy with how it was checking flags](https://github.com/akkadotnet/Alpakka/blob/ca6ff62f664bee21f764efb18de42f459c490a77/src/Channels/Akka.Streams.Channels/Internal/ChannelReaderSource.cs#L74-L92).
 - The Failure/Completion stages are less perf sensitive than the main read loop, so the code on our happy path is now smaller

Reply:
LGTM